### DISCUSSION TITLE
Need help about timezone string

### DISCUSSION THREAD

**Discussion by chrbar11** [roleoroleo/yi-hack-Allwinner-v2#606] [Q&A]
Hi, thanks for this amazing hack!

I got "Yi Outdoor 1080p RFUS" with firmware 9.0.26* and I've installed h30ga latest release (0.2.6).
It worked perfectly.

But there is a built-in YI timestamp at left top, and I'm not able to set it.

This YI timestamp works correctly when Cloud is enable, and I thought it didn't use Timezone field from "Configurations" page of YI-HACK.
But I was able to validate that it can use Timezone field value from YI-HACK.

My problem is that I don't find the good timezone string to be able to set this YI timestamp!

To show you the used Timezone field result, I've generated some snapshot with timestamp at the right bottom (snapshot.sh?res=low&watermark=yes").

If I set no Timezone value, both timestamp are the same (and wrong for my own timezone ) :
![20230325-1809](https://user-images.githubusercontent.com/45576003/227744804-361ac7e4-3ed7-4647-9b50-6a314affc12b.jpg)
_Local time  = 18:09:46 (UTC-4h)
Yi timestamp = 22:09:46 (UTC)
Snapshot timestamp = 22:09:46 (UTC)_

If I set my own Timezone value "EST5EDT,M3.2.0,M11.1.0" (America/Toronto), both timestamp are different (snapshot timestamp is correct, and YI timestamp is wrong) :
![20230325-1817](https://user-images.githubusercontent.com/45576003/227745136-418e6b9b-e55e-4bd3-824a-51f6c878101c.jpg)
_Local time  = 18:17:06 (UTC-4h)
Yi timestamp = 02:17:06 (UTC+4h, next day)
Snapshot timestamp = 18:17:06 (UTC-4h)_

If I set some Timezone value, as "HKT-8" (Asia/Hong Kong), both timestamp are the same (and correct for targeted timezone, Asia/Hong Kong) :
![20230325-1825](https://user-images.githubusercontent.com/45576003/227745306-e1455abd-4be6-48dc-89de-57ba7910cfbe.jpg)
_Local time  = 18:25:02 (UTC-4h)
Yi timestamp = 06:25:02 (UTC+8h, next day)
Snapshot timestamp = 06:25:02 (UTC+8h, next day)_

I've tried many string for my own timezone (EST5EDT, EST5, EST-4, HNE-4, EDT-4, UTC-4...) and I've tried to play with timezone values to truncate the timestamp result (as AKST9AKDT), but nothing works!

For all these test, settings from "Configurations" page of YI-HACK were:
Disable Cloud [enable]
NTPD [enable]
NTP Server [pool.ntp.org]

I've checked the firmware time :
```
root@camera2:~# date
Sat Mar 25 22:38:03 GMT 2023
root@camera2:~# date +"%Z %z"
GMT +0000
```
and it looks like firmware time is always on default UTC timezone, and has no impact on camera timestamp.

I prefer disable Cloud, and currently I've set YI-HACK to record video on SD card and send it on my NAS, thus I need that YI timestamp is the true one.

Do you have some advices to resolve this problem?

Thanks a lot for your help :)
Chris

**Comment by roleoroleo** [comment]
How did you configure this setting?
![immagine](https://user-images.githubusercontent.com/39277388/227780635-7cd34d32-eed7-44c7-aeac-80df18a3c5b9.png)

Does your cam shows the timestamp natively?

**Reply by chrbar11** [reply]
Hello,
Yes my cam shows this timestamp natively (at left top).
I didn't find settings about it on Yi App.

Cam model is Yi Outdoor 1080p RFUS (https://www.yitechnology.com/yi-outdoor-h30).
The native/current firmware version is 9.0.26.07_202104231123.
Yi App offer to update to the firmware version 11.1.26.1_202301191526 (but I didn't do it because I don't know if YI-HACK supports it).
I've used the  latest release (0.2.6) of Yi-Hack h30ga.

I don't find "Enable/disable time OSD" settings.
Is it on Yi App or on Yi-Hack Web GUI?

Thanks,
Chris

**Comment by roleoroleo** [comment]
Sorry, I added osd option after the last release.
Do you want to test a beta version?

Firmware 11 is supported for your model.

**Reply by chrbar11** [reply]
> Firmware 11 is supported for your model.

I left my camera with the native Yi firmware (9.0.26.07_202104231123)
Do you recommand to me to update it to 11.1.26* Yi firmware?

Do I have to leave the SD Card (with Yi-Hack files) in the camera during the Yi firmware update?
After the update to to 11.1.26* Yi firmware, do I have to format my SD Card and run a hack again (as the first time I apply the hack)?

**Reply by roleoroleo** [reply]
The better procedure should be:
- remove the sd card
- start the cam
- update the fw to 11
- insert again the sd card
- start the cam

**Comment by chrbar11** [comment]
Yes sure, I would like to try it.
Thanks a lot :)

**Comment by roleoroleo** [comment]
Here it is:
[h30ga_0.2.6.tar.gz](https://github.com/roleoroleo/yi-hack-Allwinner-v2/files/11079595/h30ga_0.2.6.tar.gz)

**Reply by chrbar11** [reply]
Hi roleo.

I've tried beta version with "Enable/disable time OSD" feature, but the "time OSD in the upper left corner" is still present.
Because my cam (h30ga) shows osd natively, this feature will not work for me.

I'm ok to use this native "time OSD in the upper left corner" as my one and only one timestamp on my records.
My trouble is that I can't set the good timezone on it when Cloud is disable.

My timezone is America/Toronto.

When Cloud is enable (I set the "Eastern Time" on Yi App, and "EST5EDT,M3.2.0,M11.1.0" on Yi-Hack), all timestamp (system/snapshot timestamp, and native OSD timestamp) are good.

When Cloud is disable, native OSD timestamp (upper left corner) is wrong.

I've set "Coordinated universal time" on Yi App, and "EST5EDT,M3.2.0,M11.1.0" on Yi-Hack, but the native OSD timestamp is still wrong.

Do you know which timezone string can I set on Yi-Hack, to obtain a good native OSD timestamp?

Thanks a lot for your help :)
Chris

**Comment by noman1228** [comment]
I can confirm this on:
y211ga - 0.2.6 (base 11.0.36.7_202202170022)
y291ga - 0.2.6 (base 12.0.35.8_202211241029)

Both display correct time in shell, and on config page. Neither display correct timestamp on video.

**Reply by roleoroleo** [reply]
What scenario? Cloud disabled?
Time configured inside the app? Time configured in the web page? Both?
What's your timezone string?

**Reply by noman1228** [reply]
Camera originally activated on app. 
Cloud disabled currently  POSIX string EST5EDT,M3.2.0,M11.1.0
NTP enabled on web page.
shows difference of +8 hours to current date/time.

When connected back to cloud, the time displays correct.
Either way, the time on the web page shows correct.
![snapshot](https://user-images.githubusercontent.com/163939/233669252-d83bd766-2c59-4651-93a3-23bd49342a24.jpg)
[configs.zip](https://github.com/roleoroleo/yi-hack-Allwinner-v2/files/11296958/configs.zip)

**Comment by chrbar11** [comment]
Hello,

I did many tests, I resume them below.

Model Suffix  	h30ga
Base Version 	11.1.26.1_202301191526
Firmware Version 	0.2.6 beta (with "Enable/disable time OSD" feature).

"Enable/disable time OSD" don't work for me because my cam shows OSD natively.
But I'm ok to use this native "time OSD in the upper left corner", the only one trouble is to set the good timezone on it when Cloud is disable.

There are 4 timestamp:
- "upper left corner timestamp" defined as Yi native timestamp named "time OSD" in Yi-Hack Configurations page,
- "lower right corner timestamp" defined as Yi-Hack watermark timestamp showed from snapshot,
- "system timestamp" defined as "Local Time" in Yi-Hack Status page,
- "events timestamp" defined as "Directory name" (where videos are saved on SD card) in Yi-Hack Events page.

I use https://www.timeanddate.com/time/map/ Web site to validate the UTC result.

"events timestamp" directory name is alway named with UTC-0 timezone (whatever the settings, Cloud enable/disable, "Local Time" field empty or not).

"lower right corner timestamp" and "system timestamp" results are always the same, depending of timezone string in Yi-Hack Configurations page.

"upper left corner timestamp" is always the correct time when Cloud is enable, depending of timezone value set in Yi App.

"upper left corner timestamp" is always UTC-0 time when Cloud is disable (with no timezone value in Yi-Hack Configurations page).

**With Cloud disable, "upper left corner timestamp", "lower right corner timestamp" and "system timestamp" are the same and the correct time, with the below timezone strings (in Yi-Hack Configurations page) :**
UTC-0 (to UTC-12)
UTC-00(to UTC-12)
UTC-00:00 (to UTC-12:00)

But it works for East UTC timezone only, from UTC 0 to +12 (and more).
![image](https://user-images.githubusercontent.com/45576003/233810264-0764ddc1-2b36-4623-8732-1998ec6768f4.png)

It doesn't work for West UTC timezone, from UTC -11 to -1.
![image](https://user-images.githubusercontent.com/45576003/233810269-c14f2544-fa74-4c43-900a-03ada341c303.png)

"lower right corner timestamp" and "system timestamp" are the same and the correct time (UTC less string number), but "upper left corner timestamp" is the wrong time (UTC more string number), with the below timezone strings :
UTC1
1-UTC
1
UTC+1
UTC+01
UTC+01:00

For example, with one of timezone strings above, if it's noon (12:00PM) at UTC, "lower right corner timestamp" and "system timestamp" are 11:00AM, and "upper left corner timestamp" is 13:00PM.

"upper left corner timestamp", "lower right corner timestamp" and "system timestamp" are the same and the wrong time (egal to UTC 0), when there is no timezone value (or a unkown value) in Yi-Hack Configurations page.

**Thus, I can conclude that timezone value in Yi-Hack Configurations page is used by "upper left corner timestamp".
But "upper left corner timestamp" recognizes East UTC timezone only, from UTC 0 to +12 (and more).**

My timezone is America/Toronto (UTC/GMT -4 hours) with Eastern Daylight Time or Eastern Standard Time.
My timezone string should be "EST5EDT,M3.2.0,M11.1.0".
But when I use "EST5EDT,M3.2.0,M11.1.0", "upper left corner timestamp" time is UTC/GMT +4 hours (Asia/Dubai), instead of UTC/GMT -4 hours (America/Toronto).

I've tried many timezone strings like EST5EDT, M3.2.0, M11.1.0, EST, EST5, EST-5, EST+5, EDT, EDT5, EDT-5, EDT+5... but it's always the wrong result.

For your information, with Cloud disable, all result are the same with all Timezone value set in Yi App (UTC-00:00 "Coordinated universal time" or any other).

Do you have any idea how to set West UTC timezone, from UTC -11 to -1, that "upper left corner timestamp" is going to recognize correctly?

**Reply by noman1228** [reply]
my hack-y fix so far. 
SSH into cam:
set_tz_offset -v -14400

cronttab:
# 0 * * * * set_tz_offset -v -14400

seems to work for me so far...

**Reply by noman1228** [reply]
crontab doesn't seem to be updating on reboot but manually SSH to each camera with the command works until reboot.

**Comment by chrbar11** [comment]
@noman1228 you find something, thanks a lot.

@roleoroleo Base Version and Firmware Version react differently in my side.

I did a test with two camera, same model and different Base/Firmware Version.
"set_tz_offset -v -14400" work on one, and not on the other.

Both tested camera have these common settings.

**Configurations > General**
Disable Cloud = ON
Recording without Cloud = ON
Timezone = EST5EDT,M3.2.0,M11.1.0

**Comment by chrbar11** [comment]
### TEST CAMERA 1

Hostname 	camera1
Firmware Version 	0.2.6
Base Version 	9.0.26.07_202104231123
Model Suffix 	h30ga

**Status > System**
Local Time 	Sun Apr 23 13:44:56 EDT 2023

**Events > Select time**
Date: 2023-04-23 Time: 13:00 	2023Y04M23D17H
Time: 13:44 	E4....mp4

**Snapshot**
![image](https://user-images.githubusercontent.com/45576003/233857644-1df5864b-26d0-4ad9-bd1e-22967f01efcc.png)

**Results**
Local time 2023-04-23 13:44 (America/Toronto UTC/GMT -4 hours)
- "upper left corner timestamp" = wrong (UTC/GMT +4 hours)
- "lower right corner timestamp" = ok
- "system timestamp" = ok
- "events timestamp" = ok

**SSH on camera**
```
root@camera1:~# set_tz_offset -v -14400
Setting timezone to -14400000 ms
root@camera1:~#
```

**Status > System**
Local Time 	Sun Apr 23 13:50:07 EDT 2023

**Events > Select time**
Date: 2023-04-23 Time: 13:00 	2023Y04M23D17H
Time: 13:50 	E4....mp4

**Snapshot**
![image](https://user-images.githubusercontent.com/45576003/233857962-6800865b-5aa7-4c24-8c5b-dd212bf82c6d.png)

**Result**
Local time 2023-04-23 13:50 (America/Toronto UTC/GMT -4 hours)
- "upper left corner timestamp" = ok
- "lower right corner timestamp" = ok
- "system timestamp" = ok
- "events timestamp" = ok

**Comment by chrbar11** [comment]
camera1 and camera2 came from the same Yi box (Base Version 9.0.26.07_202104231123).
Both was activated from Yi App.
Both have been hacked with Yi-Hack 0.2.6.

I did some updates on camera2:
1. update Firmware Version from 0.2.6 to 0.2.6 BETA
2. update Base Version (via Yi App) from 9.0.26.07_202104231123 to 11.1.26.1_202301191526
3. re-apply Yi-Hack 0.2.6 BETA (as the first time hack)


### TEST CAMERA 2

Hostname 	camera2
Firmware Version 	0.2.6 BETA
Base Version 	11.1.26.1_202301191526
Model Suffix 	h30ga

**Status > System**
Local Time 	Sun Apr 23 14:03:31 EDT 2023

**Events > Select time**
Date: 2023-04-23 Time: 18:00 	2023Y04M23D18H
Time: 18:03 	E4....mp4

**Snapshot**
![image](https://user-images.githubusercontent.com/45576003/233858703-607e30f8-91eb-4814-ba3d-42eed652ea32.png)

**Results**
Local time 2023-04-23 14:03 (America/Toronto UTC/GMT -4 hours)
- "upper left corner timestamp" = wrong (UTC/GMT +4 hours)
- "lower right corner timestamp" = ok
- "system timestamp" = ok
- "events timestamp" = wrong (UTC/GMT 0 hours)

**SSH on camera**
```
root@camera2:~# set_tz_offset -v -14400
root@camera2:~#
```

**Status > System**
Local Time 	Sun Apr 23 14:04:14 EDT 2023

**Events > Select time**
Date: 2023-04-23 Time: 18:00 	2023Y04M23D18H
Time: 18:04 	E4....mp4

**Snapshot**
![image](https://user-images.githubusercontent.com/45576003/233858718-092a3d37-0dcd-4d2d-94d7-4f33ec42f501.png)

**Result**
Local time 2023-04-23 14:04 (America/Toronto UTC/GMT -4 hours)
- "upper left corner timestamp" = wrong (UTC/GMT +4 hours)
- "lower right corner timestamp" = ok
- "system timestamp" = ok
- "events timestamp" = wrong (UTC/GMT 0 hours)


SSH command set_tz_offset looks like not recognized or applied.

Do you have some advices to resolve these problem and to be able to apply "set_tz_offset -v -14400"?

I don't think I can downgrade Base Version from 11.1.26.1_202301191526 to 9.0.26.* ?
Do I have to try to downgrade Firmware Version from 0.2.6 BETA to 0.2.6 ?

**Reply by chrbar11** [reply]
@roleoroleo I've "Reset to yi-hack default" and re-apply Yi-Hack 0.2.6 NOT BETA (as the first time hack) on camera2, and now SSH command set_tz_offset runs correctly:
```
root@(none):~# set_tz_offset -v -14400
Setting timezone to -14400000 ms
root@(none):~#
```

![image](https://user-images.githubusercontent.com/45576003/233864803-ae36505d-bf46-4b64-aba5-602f9beb8c84.png)

Now results are :
Local time 2023-04-23 16:30 (America/Toronto UTC/GMT -4 hours)

- "upper left corner timestamp" = ok
- "lower right corner timestamp" = ok
- "system timestamp" = ok
- "events timestamp" = wrong (UTC/GMT 0 hours)

**Comment by noman1228** [comment]
Events timestamp could be coded in firmware to save as the displayed timestamp?
I can try it locally but again, it's going to be SUPER hack-y. I'm not great but I can stumble around enough to see if it works for me.
I gotta take a camera down, make the edit and compile. I'll report back.

**Comment by roleoroleo** [comment]
There are recent fixes about this issue.
Please try this beta otherwise its too hard for me to find the problem.
[h30ga_0.2.6.tgz](https://github.com/roleoroleo/yi-hack-Allwinner-v2/files/11331786/h30ga_0.2.6.tgz)
[y211ga_0.2.6.tgz](https://github.com/roleoroleo/yi-hack-Allwinner-v2/files/11331789/y211ga_0.2.6.tgz)
[y291ga_0.2.6.tgz](https://github.com/roleoroleo/yi-hack-Allwinner-v2/files/11331790/y291ga_0.2.6.tgz)

**Reply by noman1228** [reply]
Sorry for the delay, I'll load them up this afternoon as I have two fresh y291 arriving.

**Reply by noman1228** [reply]
Both successful. The two new one's that arrived were Y21 and also don't have the issue with the public hack. I will say I LOVE the new options though. Is this the build I will get if I compile from source from the repo? I've been trying to hack together an ESP32CAM for a time-lapse camera and this blows that idea out of the water.

**Reply by roleoroleo** [reply]
> Is this the build I will get if I compile from source from the repo?

Yes, the repo is updated.

**Reply by noman1228** [reply]
Thanks again Lorenzo. Enjoy your beer!

**Reply by roleoroleo** [reply]
Thank you for your donation!

**Comment by chrbar11** [comment]
Thanks a lot @roleoroleo and @noman1228 :)

In my side, I choose to run Yi-Hack 0.2.6 (current release, not beta version) on model h30ga with base version 9.0.26*.
Yi App and all the Cloud features are disable, recording on the SD is enable and by using the crontab syntax :
`*/30 * * * * set_tz_offset -v -14400`
All timestamp are ok and evrything work great.

I'll retry last beta version on an old camera later.

**Reply by chrbar11** [reply]
@roleoroleo, I did some tests to set timezone (America/Toronto) for OSD (native upper left corner timestamp on "Yi Outdoor 1080p RFUS" with firmware 9.0.26*, Yi-Hack h30ga).

It work correctly on 0.2.6 by using crontab syntax :
*/30 * * * * set_tz_offset -v -14400

It didn't work on 0.2.6 betas (post above from Mar 27 and Apr 26) by using the same crontab syntax.

I've just done a test with the new 0.2.7, and it didn't work too (by using the same crontab syntax).

On 0.2.7 release page, it's mentioned the new feature "Add to set_tz_offset the ability to set timezone for osd".
Which settings do I have to use?

Thanks a lot for your help :)
Chris

**Reply by roleoroleo** [reply]
If the option:
![immagine](https://github.com/roleoroleo/yi-hack-Allwinner-v2/assets/39277388/a12c412d-718e-4120-af47-973537c5435a)
is on it should work automatically.
This is the code inside system.sh:
```
if [[ $(get_config TIME_OSD) == "yes" ]] ; then
    log "Enable time osd"
    # Enable time osd
    set_tz_offset -c osd -o on
    # Set timezone for time osd
    TIMEZONE=$(get_config TIMEZONE)
    TZP=$(TZ=$TIMEZONE date +%z)
    TZP_SET=$(echo ${TZP:0:1} ${TZP:1:2} ${TZP:3:2} | awk '{ print ($1$2*3600+$3*60) }')
    set_tz_offset -c tz_offset_osd -m $MODEL_SUFFIX -f $HV -v $TZP_SET
fi
```
`
set_tz_offset -c osd -o on`
activate the osd (in your case this command has no effect).
`set_tz_offset -c tz_offset_osd -m h30ga -f 9 -v $TZP_SET`
write the timezone offset in flash
where TZP_SET is the timezone in seconds (-14400 in your case).

**Reply by chrbar11** [reply]
Hi @roleoroleo, I did some tests but I can't get OSD time to work on my camera.

My camera is an Yi Outdoor 1080p RFUS.

I did a full reset (unbrick and reset pin).
I've installed the hack 0.2.7 without the Yi app.

Firmware Version 	0.2.7
Base Version 	9.0.26.07_202104231123
Model Suffix 	h30ga

I've set it step by step and I've check result at each step.

### STEP/TEST 1 (yi-hack default settings)

Timezone [empty]
Disable Cloud [down]
Recording without Cloud [down]
Enable/disable time OSD [down]
Debug log [down]

**Results**
upper left corner timestamp = UTC -/+0
lower right corner timestamp = UTC -/+0
system timestamp = UTC -/+0
events timestamp = (not avalaible)

![snapshot1](https://github.com/roleoroleo/yi-hack-Allwinner-v2/assets/45576003/59574b5f-687d-4b31-85a8-53bcb2955a5f)

### STEP/TEST 2 (disable Cloud and enable recording without Cloud)

Timezone [empty]
Disable Cloud [**up**]
Recording without Cloud [**up**]
Enable/disable time OSD [down]
Debug log [**up**]

**Results**
upper left corner timestamp = UTC -/+0
lower right corner timestamp = UTC -/+0
system timestamp = UTC -/+0
events timestamp = UTC -/+0

![snapshot2](https://github.com/roleoroleo/yi-hack-Allwinner-v2/assets/45576003/dc634a83-91aa-486a-bb4e-023bf56e0bbb)

### STEP/TEST 3 (add timezone)

Timezone [**EST5EDT,M3.2.0,M11.1.0**]
Disable Cloud [up]
Recording without Cloud [up]
Enable/disable time OSD [down]
Debug log [up]

**Results**
upper left corner timestamp = **wrong (UTC +4)**
lower right corner timestamp = ok (UTC -5)
system timestamp = ok (UTC -5)
events timestamp = ok (UTC -5)

![snapshot3](https://github.com/roleoroleo/yi-hack-Allwinner-v2/assets/45576003/3930d06e-1d83-4caa-8d28-b086dd77f89a)

### STEP/TEST 4 (enable time OSD)

Timezone [EST5EDT,M3.2.0,M11.1.0]
Disable Cloud [up]
Recording without Cloud [up]
Enable/disable time OSD [**up**]
Debug log [up]

**Results**
upper left corner timestamp = **wrong (UTC +4)**
lower right corner timestamp = ok (UTC -5)
system timestamp = ok (UTC -5)
events timestamp = ok (UTC -5)

![snapshot4](https://github.com/roleoroleo/yi-hack-Allwinner-v2/assets/45576003/21dfc92c-b3cc-4e8c-b34c-d82b6df39756)

### Set OSD time manualy

The command line "set_tz_offset -v -14400" doesn't work (via SSH or via crontab) under 0.2.7 version (and under 0.2.6 betas version too)
Is it possible to allow it under 0.2.7?

Thanks a lot for your help :)
Chris

**Reply by chrbar11** [reply]
Hi @roleoroleo, I saw that you made two new releases, 0.2.8 and 0.2.9.

Does any of them fix "time OSD" (upper left corner timestamp) which is natively displayed on my camera and has wrong time zone when Cloud is disabled (this issue remained with 0.2.7 on h30ga and Yi Outdoor 1080p RFUS)?

If not, does any of them allow running "set_tz_offset -v -14400" from Crontab (0.2.6beta and 0.2.7 didn't).

Thank you so much :)
Chris

**Reply by roleoroleo** [reply]
h30ga with fw 9 should work.
fw 12 needs another release.
Do tyou want to try a beta?

What's the problem with crontab?

**Reply by chrbar11** [reply]
> What's the problem with crontab?

Hi @roleoroleo, when I add
*/30 * * * * set_tz_offset -v -14400
in the crontab (Yi-Hack GUI), it's correctly applied (after 30mn) under 0.2.6, but nothing under 0.2.6beta and 0.2.7.

I can see the difference via SSH.

Under 0.2.6 :
```
root@(none):~# set_tz_offset -v -14400
Setting timezone to -14400000 ms
root@(none):~#
```
And under 0.2.6beta and 0.2.7 :
```
root@(none):~# set_tz_offset -v -14400
root@(none):~#
```

The command seems to be known, but maybe not the value (or whatever).

If nobody has reported this issues, it could concern h30ga only.

**Reply by roleoroleo** [reply]
Try to run the program with --help and check other required options.
Probably you miss -c option.
